src/script/lua_api/l_particleparams.h

```diff
@@ -0,0 +1,199 @@
+
+This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation; either version 2.1 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#pragma once
+#include "lua_api/l_particles.h"
+#include "lua_api/l_object.h"
+#include "lua_api/l_internal.h"
+#include "common/c_converter.h"
+#include "common/c_content.h"
+#include "server.h"
+#include "particles.h"
+
+namespace LuaParticleParams {
+	using namespace ParticleParamTypes;
+
+	template<typename T>
+	inline void readNumericLuaValue(lua_State* L, T& ret) {
+		if (lua_isnil(L,-1)) return;
+		if (std::is_integral<T>())
+			ret = lua_tointeger(L, -1);
+		else
+			ret = lua_tonumber(L, -1);
+	}
+
+	template <typename T>
+	inline void readNumericLuaValue(lua_State* L, NumericParameter<T>& ret) {
+		readNumericLuaValue<T>(L, ret.val);
+	}
+
+	// these are unfortunately necessary as C++ intentionally disallows function template
+	// specialization and there's no way to make template overloads reliably resolve correctly
+	inline void readLuaValue(lua_State* L, f32Parameter& ret) { readNumericLuaValue(L, ret); }
+	inline void readLuaValue(lua_State* L, f32& ret)          { readNumericLuaValue(L, ret); }
+	inline void readLuaValue(lua_State* L, u16& ret)          { readNumericLuaValue(L, ret); }
+	inline void readLuaValue(lua_State* L, u8& ret)           { readNumericLuaValue(L, ret); }
+
+	inline void readLuaValue(lua_State* L, v3fParameter& ret) {
+		if (lua_isnumber(L, -1)) { // shortcut for uniform vectors
+			auto n = lua_tonumber(L, -1);
+			ret = v3fParameter(n,n,n);
+		} else {
+			ret = (v3fParameter)check_v3f(L, -1);
+		}
+	}
+
+	inline void readLuaValue(lua_State* L, v2fParameter& ret) {
+		if (lua_isnumber(L, -1)) { // shortcut for uniform vectors
+			auto n = lua_tonumber(L, -1);
+			ret = v2fParameter(n,n);
+		} else {
+			ret = (v2fParameter)check_v2f(L, -1);
+		}
+	}
+
+	inline void readLuaValue(lua_State* L, TweenStyle& ret) {
+		const char* const opts[] = { "fwd", "rev", "pulse", "flicker", NULL };
+		const TweenStyle optmap[] = {
+			TweenStyle::fwd,
+			TweenStyle::rev,
+			TweenStyle::pulse,
+			TweenStyle::flicker,
+		};
+		static_assert(
+				(sizeof opts / sizeof opts[0])-1 ==
+				(sizeof optmap / sizeof optmap[0]),
+				"option maps not synced");
+
+		if (lua_isstring(L, -1)) {
+			size_t v = luaL_checkoption(L, -1, opts[0], opts);
+			ret = optmap[v];
+		}
```

Comment:
we have `string_to_enum` (common/c_content.h) for this

Reply:
oh thank you!